### PR TITLE
get comments by id

### PR DESCRIPTION
**Pull request by majorhaji**
ticket 6

### CODE REVIEW

File: __tests__/app.test.js [hunk 1]
```diff
@@ -109,6 +109,50 @@ describe("get article by id", () => {
...
+    return request(app)
+      .get("/api/articles/1/comments")
+      .expect(200)
+      .then(({ body }) => {
+        const comments = body.comments;
+        expect(comments).toBeSorted({
+          descending: true,
+          key: "created_at",
+        });
+      });
```
Comment (Poonam-raj) [inline __tests__/app.test.js:124]: nice sorting check!

File: __tests__/app.test.js [hunk 2]
```diff
@@ -109,6 +109,50 @@ describe("get article by id", () => {
       });
   });
 });
+
+describe("get comments by article id", () => {
+  it("200: returns comments for given id", () => {
```
Comment (Poonam-raj) [inline __tests__/app.test.js]: i think this test description and the description on line 135 should be swapped to make more sense for what's being asserted

File: __tests__/app.test.js [hunk 3]
```diff
@@ -109,6 +109,50 @@ describe("get article by id", () => {
...
+          key: "created_at",
+        });
+      });
+  });
+  it("404: returns message if article has no comments", () => {
+    return request(app)
+      .get("/api/articles/10/comments")
+      .expect(404)
+      .then(({ text }) => {
+        expect(text).toBe("Path not found");
```
Comment (Poonam-raj) [inline __tests__/app.test.js]: When an article is valid and exists in our database but doesn't have any comments we can actually consider that a happy path. The user hasn't done anything wrong, we haven't done anything wrong, no one's commented on the article is all. 

So for this case I'd expect an empty comments array and a 200 status to come back

Comment (Poonam-raj) [reply __tests__/app.test.js]: sidenote: really glad you even spotted that this was a scenario

File: __tests__/app.test.js [hunk 4]
```diff
@@ -109,6 +109,50 @@ describe("get article by id", () => {
...
+              votes: expect.any(Number),
+              created_at: expect.any(String),
+              author: expect.any(String),
+              body: expect.any(String),
+            })
+          );
+        });
+      });
+  });
+});
```
Comment (Poonam-raj) [inline __tests__/app.test.js:171]: Can't see it here, we'd need a 400 invalid ID and a 404 non-existent ID set of tests for this endpoint still

**Review by Poonam-raj — CHANGES_REQUESTED** [review]
Hi Yusuf, I want you to amend your test suite for this endpoint and send another PR back to us so we can give it a once over. Especially that happy path of a review not having an comments but still being valid. 

read through the comments carefully and throw any qs my way if need be

File: __tests__/app.test.js [hunk 5]
```diff
@@ -109,6 +109,57 @@ describe("get article by id", () => {
...
+      .get("/api/articles/404044/comments")
+      .expect(404)
+      .then(({ text }) => {
+        expect(text).toBe("Path not found");
+      });
+  });
+
+  it("400: returns message if bad request", () => {
+    return request(app)
+      .get("/api/articles/banana")
```
Comment (Pipe-Smoking-Rabbit) [inline __tests__/app.test.js:137]: `banana` is an excellent choice of string

File: __tests__/app.test.js [hunk 6]
```diff
@@ -109,6 +109,57 @@ describe("get article by id", () => {
...
+        const comments = body.comments;
+        expect(comments).toBeSorted({
+          descending: true,
+          key: "created_at",
+        });
+      });
+  });
+  it("404: returns message if path not found", () => {
+    return request(app)
+      .get("/api/articles/404044/comments")
```
Comment (Pipe-Smoking-Rabbit) [inline __tests__/app.test.js:128]: just reconsider the error `text` on this error - the error here is that the article hasn't been found in the database - Path not found relates to a bad path i.e. `/api/invalid-endpoint/` which is something we've already tested for

File: models/models.articles.js [hunk 7]
```diff
@@ -34,3 +34,20 @@ exports.selectArticleById = (id) => {
...
+  return db
+    .query(
+      `SELECT comment_id, comments.votes, comments.created_at, comments.author, comments.body FROM articles JOIN comments ON articles.article_id = comments.article_id WHERE articles.article_id=$1 ORDER BY created_at DESC;`,
+      [id]
+    )
+    .then(({ rows, rowCount }) => {
+      if (rowCount === 0) {
+        return Promise.reject({
+          status: 404,
+          msg: "Path not found",
```
Comment (Pipe-Smoking-Rabbit) [inline models/models.articles.js]: we don't need this `if ( rowCount === 0) return Promise.reject` section - some articles will have no comments and that is okay - we don't want to throw errors for that 

You'll have to check that the `article_id` is in the database another way  ( you can make use of another model you made on a previous ticket `selectArticleByID` which is already set to throw an error if the `article_id` is not found )

File: controllers/controllers.articles.js [hunk 8]
```diff
@@ -23,3 +24,17 @@ exports.getArticleById = (req, res, next) => {
       next(err);
     });
 };
+
+exports.getCommentsByArticleId = (req, res, next) => {
+  const { article_id } = req.params;
+
+  selectCommentsByArticleId(article_id)
+    .then((rows) => {
```
Comment (Pipe-Smoking-Rabbit) [inline controllers/controllers.articles.js:34]: here is where you can make use of `two` model functions, including your `selectArticleById` to check that the `article_id` is in the database

**Review by Pipe-Smoking-Rabbit — CHANGES_REQUESTED** [review]
Nice work Yusuf! It's really close but still throwing an error if there are no comments on a `valid` article, I've left comments for how you can get around this - if you're unsure about how to implement it then put in an `nchelp` and we can go through it together 

Please do another PR request for this ticket once it's all sorted

File: __tests__/app.test.js (hunk 6 above)
Comment (Poonam-raj) [reply __tests__/app.test.js:128]: Again I agree that the message being sent back here is not correct. This test should be looking more like:
``` 
expect(msg).toBe("Article not found")
```

Comment (Poonam-raj) [reply __tests__/app.test.js:128]: I think your selectArticleById needs amending as well as tests relating to this 404 article not found sad path

**Review by Poonam-raj — APPROVED** [review]
Hey, one thing from me:
- Amend your 404 article not found test, it should be expecting something more clear and a better key name

Once you've amended that and gotten everything passing again you're free to merge this PR

**Review by Poonam-raj — APPROVED** [review]
Excellent stuff, this is exactly what I was hoping for 🚀  Feel free to merge

**Pull request merged**